service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java

```diff
@@ -297,13 +332,58 @@ public TelemetrySender downstreamTelemetrySender(
      */
     @Qualifier(EventConstants.EVENT_ENDPOINT)
     @Bean
+    @ConditionalOnAmqpMessaging
     @Scope("prototype")
     public EventSender downstreamEventSender(
             final SendMessageSampler.Factory samplerFactory,
             final ProtocolAdapterProperties adapterConfig) {
         return new ProtonBasedDownstreamSender(downstreamConnection(), samplerFactory, adapterConfig);
     }
 
+    /**
+     * Exposes a factory for creating producers for sending downstream messages via the Kafka cluster.
+     *
+     * @return The factory.
+     */
+    @Bean
+    @Scope("prototype")
+    @ConditionalOnKafkaMessaging
+    public CachingKafkaProducerFactory<String, Buffer> kafkaProducerFactory() {
+        return CachingKafkaProducerFactory.sharedProducerFactory(vertx());
+    }
+
+    /**
+     * Exposes a client for sending telemetry messages via <em>Kafka</em> as a Spring bean.
+     *
+     * @return The client.
+     * @param kafkaProducerFactory The producer factory to use.
+     * @param kafkaProducerConfig The producer configuration to use.
+     */
+    @Bean
+    @ConditionalOnKafkaMessaging
+    @Scope("prototype")
+    public TelemetrySender downstreamTelemetryKafkaSender(
+            final CachingKafkaProducerFactory<String, Buffer> kafkaProducerFactory,
+            final KafkaProducerConfigProperties kafkaProducerConfig) {
+        return new KafkaBasedTelemetrySender(kafkaProducerFactory, kafkaProducerConfig, getTracer());
+    }
+
+    /**
+     * Exposes a client for sending events via <em>Kafka</em> as a Spring bean.
+     *
+     * @return The client.
+     * @param kafkaProducerFactory The producer factory to use.
+     * @param kafkaProducerConfig The producer configuration to use.
+     */
+    @Bean
+    @ConditionalOnKafkaMessaging
+    @Scope("prototype")
+    public EventSender downstreamEventKafkaSender(
+            final CachingKafkaProducerFactory<String, Buffer> kafkaProducerFactory,
+            final KafkaProducerConfigProperties kafkaProducerConfig) {
+        return new KafkaBasedEventSender(kafkaProducerFactory, kafkaProducerConfig, getTracer());
```

Comment:
both the event sender  and the telemetry sender seem to use the same configuration, or am I mistaken? If that is the case, what advantage does it have to create two separate senders? is there a reason why they shouldn't use the same KafkaProducer instance (and the same connection to the Kafka cluster)?

Reply:
These are very valid questions. I just had no idea where I could explain the reasoning behind it.

In general, a Kafka client does not open a single connection but rather connects to the brokers to whose partitions it sends messages, i.e. between 1 and the cluster size. Since the topics for events and telemetry do not have to be on the same brokers, on average there are not many more connections through the separate sender instances.

I see it as an advantage that the "rare but important" events do not have to wait in the same queue as the much more frequent telemetry messages. Since the messages are collected in batches before sending and are sent in sequence with retries and rather generous timeouts (the default timeout for sending is [2 minutes](https://kafka.apache.org/documentation/#delivery.timeout.ms)), I expect a lower latency for events under load. 
I would also like us to have the possibility to decide later that both types of messages should be configured differently.